src/support/strongly_connected_components.h

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright 2024 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef wasm_support_strongly_connected_components_h
+#define wasm_support_strongly_connected_components_h
+
+#include <cassert>
+#include <optional>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#include <iostream>
+
+namespace wasm {
+
+// A CRTP utility implementing Tarjan's Strongly Connected Component algorithm
+// (https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm)
+// in terms of iterators. Given the beginning and end iterators over the
+// elements in a graph an implementation of `pushChildren` in the CRTP subclass
+// that pushes an element's children, provides an iterable over the SCCs, each
+// of which is an iterable over the elements in the SCC. All implemented
+// iterators are input iterators that mutate the underlying state, so this
+// utility can only be used for single-pass algorithms.
+template<typename It, typename Class> struct SCCs {
+  SCCs(It inputIt, It inputEnd) : inputIt(inputIt), inputEnd(inputEnd) {}
+
+private:
+  using T = typename It::value_type;
+
+  // The iterators over the graph we are calculating the SCCs for.
+  It inputIt;
+  It inputEnd;
+
+  // Stack of pending elements to visit, used instead of a recursive visitor.
+  struct WorkItem {
+    T item;
+    std::optional<T> parent = std::nullopt;
+    bool processedChildren = false;
+  };
+  std::vector<WorkItem> workStack;
+  // The Tarjan's algorithm stack. Similar to a DFS stack, but elements are only
+  // popped off once they are committed to a strongly connected component.
+  // Elements stay on the stack after they are visited iff they have a back edge
+  // to an element earlier in the stack.
+  std::vector<T> stack;
+  std::unordered_set<T> stackElems;
+  // Indices assigned to each element in the order they are visited.
+  std::unordered_map<T, size_t> indices;
+  // The smallest index of the elements reachable from each element through
+  // its subtree.
+  std::unordered_map<T, size_t> minReachable;
```

Comment:
Would these three perhaps be more efficient as a single map of a struct like this?
```cpp
struct Info {
  size_t index;
  size_t lowlink;
  bool onStack;
};
std::unordered_map<T, Info> infoMap; 
```
(Also those names match the names in the paper.)